include/gabia/secure_socket.hpp

```diff
@@ -0,0 +1,302 @@
+//
+// Copyright (c) 2017 Damian Jarek (damian dot jarek93 at gmail dot com)
+//
+// Distributed under the Boost Software License, Version 1.0. (See accompanying
+// file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+//
+// Official repository: https://github.com/djarek/gabia
+//
+
+#ifndef GABIA_SECURE_SOCKET_HPP
+#define GABIA_SECURE_SOCKET_HPP
+
+#include <gabia/bytes.hpp>
+#include <gabia/crypto/aead.hpp>
+
+#include <boost/asio.hpp>
+#include <boost/beast.hpp>
+#include <boost/endian/arithmetic.hpp>
+
+#define CORO_YIELD BOOST_ASIO_CORO_YIELD
+#define CORO_REENTER BOOST_ASIO_CORO_REENTER
+
+namespace gabia {
+
+namespace asio = boost::asio;
+namespace beast = boost::beast;
+
+template <typename Stream>
+class secure_socket {
+public:
+    using next_layer_type = Stream;
+
+    constexpr static auto max_payload_length = 1024;
+
+    secure_socket(asio::io_service& service) : next_layer_stream{service} {}
+
+    secure_socket(next_layer_type&& next_layer_stream,
+                  crypto::aead_secrets secrets)
+        : next_layer_stream{std::move(next_layer_stream)}, secrets{secrets} {}
+
+    auto& next_layer() { return next_layer_stream; }
+
+    const auto& next_layer() const { return next_layer_stream; }
+
+    auto& lowest_layer() { return next_layer().lowest_layer(); }
+
+    const auto& lowest_layer() const { return next_layer().lowest_layer(); }
+
+    auto& get_io_service() { return lowest_layer().get_io_service(); }
+
+    bool is_open() const { return next_layer().is_open(); }
+
+    void close() { next_layer().close(); }
+
+    template <typename ConstBufferSequence>
+    std::size_t write_some(const ConstBufferSequence& sequence,
+                           beast::error_code& error) {
+        error = {};
+        std::vector<gsl::byte> pdu{};
+        encrypt_pdu(pdu, sequence);
+        asio::write(next_layer(), asio::buffer(pdu), error);
+        if (error) {
+            return 0;
+        }
+
+        return asio::buffer_size(sequence);
```

Comment:
Isn't the ```size_t n``` argument of the completion handler intended to inform the caller how many bytes from the Sequence were actually transferred from the sequence, rather than "on the wire"?